src/main/java/mike/gui/Line.java

```diff
@@ -5,65 +5,170 @@
 import java.awt.Dimension;
 import java.awt.Graphics;
 import java.awt.Graphics2D;
+import java.awt.Point;
+import java.awt.RenderingHints;
+import java.awt.Stroke;
 import java.awt.geom.Line2D;
+import java.awt.geom.Path2D;
+
 import javax.swing.JComponent;
 import javax.swing.JLabel;
 
+import mike.datastructures.Relationship.Type;
+
 public class Line extends JComponent {
 	private static final long serialVersionUID = 1L;
-	double x1, y1, x2, y2;
-	JLabel classOne, classTwo;
-	Line(double x1, double y1, double x2, double y2, JLabel classOne, JLabel classTwo) {
+	private double x1, y1, x2, y2;
+	private JLabel classOne, classTwo;
+	private boolean toSelf;
+	private Type type;
+
+	Line(JLabel L1, JLabel L2, Type type) {
 		super();
-		
-		this.x1 = x1;
-		this.y1 = y1;
-		this.x2 = x2;
-		this.y2 = y2;
-		this.classOne = classOne;
-		this.classTwo = classTwo;
-		
+		this.type = type;
+		update(L1, L2);
+
 		this.setVisible(true);
-		this.setPreferredSize(new Dimension (782, 725));
+		this.setPreferredSize(new Dimension(classOne.getParent().getWidth(), classOne.getParent().getHeight()));
 	}
-	
-	public void setNewPoints(double newx1, double newy1, double newx2, double newy2) {
-		x1 = newx1;
-		y1 = newy1;
-		x2 = newx2;
-		y2 = newy2;
+
+	public void update(JLabel L1, JLabel L2) {
+		// (a,b) = L1 center
+		double a = L1.getLocation().x + L1.getSize().width / 2;
+		double b = L1.getLocation().y + L1.getSize().height / 2;
+
+		// (c,d) = L2 center
+		double c = L2.getLocation().x + L2.getSize().width / 2;
+		double d = L2.getLocation().y + L2.getSize().height / 2;
+
+		double[] centers1 = { a, b, c, d };
+		Point p1 = getEdgeIntersectionPoint(L1.getSize(), L1.getLocation(), centers1);
+		double[] centers2 = { c, d, a, b };
+		Point p2 = getEdgeIntersectionPoint(L2.getSize(), L2.getLocation(), centers2);
+
+		x1 = p1.getX();
+		y1 = p1.getY();
+		x2 = p2.getX();
+		y2 = p2.getY();
+		classOne = L1;
+		classTwo = L2;

```

Comment:
Do we need to do this?  Relationships can't change classes, and since this function gets called so much when moving a class it would help just a little bit.